src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -313,41 +313,35 @@ namespace stan {
     validate_non_void_arg_f;
 
     void set_void_function:: operator()(const expr_type& return_type,
-                                        var_origin& origin, bool& pass,
+                                        scope& var_scope, bool& pass,
                                         std::ostream& error_msgs) const {
       if (return_type.is_void() && return_type.num_dims() > 0) {
         error_msgs << "Void return type may not have dimensions declared."
                    << std::endl;
         pass = false;
         return;
       }
-      origin = return_type.is_void()
-        ? var_origin(void_function_argument_origin)
-        : var_origin(function_argument_origin);
+      var_scope = return_type.is_void()
+        ? scope(void_function_argument_origin)
+        : scope(function_argument_origin);
       pass = true;
     }
     boost::phoenix::function<set_void_function> set_void_function_f;
 
     void set_allows_sampling_origin::operator()(const std::string& identifier,
-                                                bool& allow_sampling,
-                                                var_origin& origin) const {
-      bool is_void_function_origin
-        = (origin.program_block_ == void_function_argument_origin);
+                                                scope& var_scope) const {
       if (ends_with("_lp", identifier)) {
-        allow_sampling = true;
-        origin = is_void_function_origin
-          ? var_origin(void_function_argument_origin_lp)
-          : var_origin(function_argument_origin_lp);
+        var_scope = var_scope.void_fun()
+          ? scope(void_function_argument_origin_lp)
+          : scope(function_argument_origin_lp);
       } else if (ends_with("_rng", identifier)) {
-        allow_sampling = false;
-        origin = is_void_function_origin
-          ? var_origin(void_function_argument_origin_rng)
-          : var_origin(function_argument_origin_rng);
+        var_scope = var_scope.void_fun()
+          ? scope(void_function_argument_origin_rng)
+          : scope(function_argument_origin_rng);
```

Comment:
You can move the ternary operator inside `scope()` --- your call.  There are a few cases of this.